components/tidb_query_vec_executors/src/table_scan_executor.rs

```diff
@@ -291,21 +295,38 @@ impl ScanExecutorImpl for TableScanExecutorImpl {
     ) -> Result<()> {
         use tidb_query_datatype::codec::{datum, table};
 
-        check_record_key(&key)?;
+        table::check_record_key(&key)?;
         let columns_len = self.schema.len();
         let mut decoded_columns = 0;
 
+        // NOTE: `primary_column_ids` is mutual exclusive toward `pk_handle`.
+        // If the column contains a pk_handle flag, we should extract the value of the column from the key.
         if !self.handle_indices.is_empty() {
-            let handle_id = table::decode_handle(key)?;
+            let handle = table::decode_int_handle(key)?;
+
             for handle_index in &self.handle_indices {
                 // TODO: We should avoid calling `push_int` repeatedly. Instead we should specialize
                 // a `&mut Vec` first. However it is hard to program due to lifetime restriction.
-                columns[*handle_index]
-                    .mut_decoded()
-                    .push_int(Some(handle_id));
+                columns[*handle_index].mut_decoded().push_int(Some(handle));
                 decoded_columns += 1;
                 self.is_column_filled[*handle_index] = true;
             }
+        } else if !self.primary_column_ids.is_empty() {
+            // Otherwise, if `primary_column_ids` is not empty, we try to extract the values of the columns from the common handle.
+            let mut handle = table::decode_common_handle(key)?;
+            for primary_id in self.primary_column_ids.iter() {
+                let index = self.column_id_index.get(primary_id);
+                if let Some(&index) = index {
+                    if !self.is_column_filled[index] {
```

Comment:
Maybe we should return error instead of continue in this case. It means that there are duplicated column id in `primary_column_ids`, which should not happen in normal cases.

Reply:
It never happens, duplicated column name is not allowed in index.
I think it's ok to just remove the if check.

Reply:
done.